### PR TITLE
Add documentation and usability for prepared parameters

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/7776

related to https://github.com/apache/arrow-datafusion/issues/7776 and https://github.com/apache/arrow-datafusion/issues/4539

## Rationale for this change
While reviewing https://github.com/apache/arrow-datafusion/issues/7776,  I noticed that the parameter values were not well documented. So I started trying to write documentation examples and in so doing found several usability papercuts  
 that I I also fixed.

## What changes are included in this PR?
1. Doc comments for `LogicalPlan::with_param_values` and YYY with examples
2. Add `placeholder` to `expr_fn` function to more easily construct placeholders
3. Move `infer_placeholder_values` to Expr, and make it public (so it can be used outside of the context of a SQL statement)
4. Change `LogicalPlan::with_param_values` to also substitute parameter values for non `PREPARE` statements as well

## Are these changes tested?
Yes, new tests / doc comments

## Are there any user-facing changes?
Better docs 

Note it is still not possible to use PREPARE / EXCUTE from SQL but I think this PR completes the backend code to make it work

### CODE REVIEW

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -1030,6 +1034,49 @@ impl Expr {
     pub fn contains_outer(&self) -> bool {
         !find_out_reference_exprs(self).is_empty()
     }
+
+    /// Find all [`Expr::Placeholder`] in anthis, and try
+    /// to infer their [`DataType`] from the context of their use.
+    pub fn infer_placeholder_types(self, schema: &DFSchema) -> Result<Expr> {
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:1043]: This was moved from the `datafusion-sql` module as it is not SQL specific, but generic to expressions

**Comment by alamb** [comment]
cc @NGA-TRAN

File: datafusion/core/src/dataframe.rs [hunk 2]
```diff
@@ -1218,7 +1218,42 @@ impl DataFrame {
...
+    ///    "| a |",
+    ///    "+---+",
+    ///    "| 1 |",
+    ///    "+---+",
+    ///  ],
+    ///  &results
+    /// );
+    /// # Ok(())
+    /// # }
+    /// ```
```
Comment (NGA-TRAN) [inline datafusion/core/src/dataframe.rs:1256]: ❤️

File: datafusion/expr/src/expr_fn.rs [hunk 3]
```diff
@@ -80,6 +80,22 @@ pub fn ident(name: impl Into<String>) -> Expr {
     Expr::Column(Column::from_name(name))
 }
 
+// Create placeholder value that will be filled in (such as `$1`):
+///
+/// For example:
+///
+/// ```rust
+/// # use datafusion_expr::{placeholder};
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply datafusion/expr/src/expr_fn.rs:90]: ❤️

File: datafusion/sql/tests/sql_integration.rs [hunk 4]
```diff
@@ -3684,6 +3684,19 @@ fn test_prepare_statement_should_infer_types() {
     assert_eq!(actual_types, expected_types);
 }
 
+#[test]
+fn test_non_prepare_statement_should_infer_types() {
+    // Non prepared statements (like SELECT) should also have their parameter types inferred
+    let sql = "SELECT 1 + $1";
```
Comment (NGA-TRAN) [inline datafusion/sql/tests/sql_integration.rs:3690]: Nice

**Review by NGA-TRAN — APPROVED** [review]
Thanks so much @alamb for the cleanup and document work

**Review by avantgardnerio — APPROVED** [review]
Seems minor and innocuous, and adds a test :+1:

**Review by jackwener — APPROVED** [review]

File: datafusion/expr/src/expr.rs [hunk 5]
```diff
@@ -599,10 +600,13 @@ impl InSubquery {
...
 
-/// Placeholder
+/// Placeholder, representing bind parameter values such as `$1`.
+///
+/// The type of these parameters is inferred using [`Expr::infer_placeholder_types`]
+/// or can be specified directly using `PREPARE` statements.
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
 pub struct Placeholder {
-    /// The identifier of the parameter (e.g, $1 or $foo)
+    /// The identifier of the parameter, including the leading `$` (e.g, `"$1"` or `"$foo'`)
```
Comment (jackwener) [inline datafusion/expr/src/expr.rs:609]:
```suggestion
    /// The identifier of the parameter, including the leading `$` (e.g, `"$1"` or `"$foo"`)
```

**Review by jackwener — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @jackwener  and @avantgardnerio

**Pull request merged**